### PR TITLE
[MLIR] Add SizeOp conversion from ONNX dialect to Krnl dialect

### PR DESCRIPTION
**Pull request by pashu123**
Added ONNXSizeOp conversion from ONNX dialect to Krnl dialect. This op is added as a part of --convert-onnx-to-krnl pass.

Signed-off-by: Prashant Kumar <pk5561@gmail.com>

### CODE REVIEW

File: src/Conversion/ONNXToKrnl/Tensor/SizeOf.cpp [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    if (hasAllConstantDimensions(memRefType))
+      alloc = insertAllocAndDealloc(memRefType, loc, rewriter, insertDealloc);
+    else
+      alloc = insertAllocAndDealloc(
+          memRefType, loc, rewriter, insertDealloc, {resultOperand});
+
+    // The minimum number of elements in the input tensor has to be 1.
+    unsigned noElements = 1;
+    for (unsigned i = 0; i < dataShape.size(); i++)
+      noElements *= dataShape[i];
```
Comment (tungld) [inline src/Conversion/ONNXToKrnl/Tensor/SizeOf.cpp:44]: Could you please handle the situation where dimensions in the `data`'s shape are unknown? E.g. `tensor<?x?xf32>`. In that situation, `dataShape[i] = -1` and we need to get the dimension via DimOp.

File: test/mlir/onnx/onnx_lowering.mlir [hunk 2]
```diff
@@ -2132,3 +2132,17 @@ func @cast_lowering_f64f32_10(%arg0: tensor<10xf64>) -> tensor<*xf32> {
...
+func @test_sizeof(%arg0: tensor<2x2xf32>) -> tensor<i64> {
+  %1 = "onnx.Size"(%arg0) : (tensor<2x2xf32>) -> tensor<i64>
+  "std.return"(%1) : (tensor<i64>) -> ()
+
+  // CHECK-LABEL: test_sizeof
+  // CHECK-NEXT: [[RES:%.+]] = alloc() : memref<i64>
+  // CHECK: [[SIZE:%.+]] = constant 4 : i64
+  // CHECK: affine.store [[SIZE]], [[RES]][] : memref<i64>
+  // CHECK: return [[RES]] : memref<i64>
+
```
Comment (tungld) [inline test/mlir/onnx/onnx_lowering.mlir:2147]: Please add a test for unknown dimensions? e.g. `tensor<?x?xf32>`. Thanks!

File: src/Conversion/ONNXToKrnl/Tensor/SizeOf.cpp [hunk 3]
```diff
@@ -0,0 +1,57 @@
+//===---------------- SizeOf.cpp - Lowering Gather_nd Op
```
Comment (tungld) [inline src/Conversion/ONNXToKrnl/Tensor/SizeOf.cpp:1]: Gather_nd => SizeOf

**Review by tungld — CHANGES_REQUESTED** [review]
@prasanthpul Thank you for submitting a patch! Could you please add more codes to handle the situation where we have unknown dimensions?

File: src/Conversion/ONNXToKrnl/CMakeLists.txt [hunk 4]
```diff
@@ -22,6 +22,7 @@ add_library(OMONNXToKrnl
         Tensor/Constant.cpp
         Tensor/Concat.cpp
         Tensor/Split.cpp
+        Tensor/SizeOf.cpp
```
Comment (tungld) [inline src/Conversion/ONNXToKrnl/CMakeLists.txt:25]: Please use `Size.cpp` to make it consistent with other ops.

**Review by tungld — CHANGES_REQUESTED** [review]

**Comment by pashu123** [comment]
> @prasanthpul Thank you for submitting a patch! Could you please add more codes to handle the situation where we have unknown dimensions?

Hi @tungld you have referenced wrong person.

File: test/mlir/onnx/onnx_lowering.mlir (hunk 2 above)
Comment (pashu123) [reply test/mlir/onnx/onnx_lowering.mlir:2157]: @tungld  Added, please check.

Comment (tungld) [reply test/mlir/onnx/onnx_lowering.mlir:2157]: Thank you! 

I think we can make the emitted code better by computing the constant part at the compilation time. For example, with tensor<2x?x2xf32>, we can compute 2x2 at the compilation time, emit a constant of value 4, and emit a MulIOp for multiplication of the first dim and the constant. Something like this:
```mlir
  // Accumulate static dimensions first.
  int64_t staticNumElement = 1;
  bool allStaticDimensions = true;
  for (unsigned i = 0; i < dataShape.size(); i++) {
    if (dataShape[i] != -1)
      staticNumElement *= dataShape[i];
    else
      allStaticDimensions = false;
  }
  // Accumulate the remaining dimensions that are unknown.
  Value noElements =
      emitConstantOp(rewriter, loc, memRefType.getElementType(), staticNumElement);
  if (!allStaticDimensions) {
    for (unsigned i = 0; i < dataShape.size(); i++) {
      if (dataShape[i] == -1) {
         Value index = rewriter.create<DimOp>(loc, data, i);
         Value dim = rewriter.create<IndexCastOp>(
             loc, index, memRefType.getElementType());
         noElements = rewriter.create<MulIOp>(loc, noElements, dim);
      }
    }
  }
```

Comment (pashu123) [reply test/mlir/onnx/onnx_lowering.mlir:2157]: @tungld Thanks, this looks good to me.

Comment (pashu123) [reply test/mlir/onnx/onnx_lowering.mlir:2153]: Hi @tungld  Could you review further? If there is no changes could you approve?

**Review by chentong319 — APPROVED** [review]
LGTM

**Comment by chentong319** [comment]
@pashu123 Do you have plan to lower ShapeOp to krnl?

**Comment by pashu123** [comment]
> @pashu123 Do you have plan to lower ShapeOp to krnl?

Yes will do that. Just wanted to know do you have meetings where things related to ONNX-MLIR are discussed. It will be helpful to know future plans.

**Comment by chentong319** [comment]
> > @pashu123 Do you have plan to lower ShapeOp to krnl?
> 
> Yes will do that. Just wanted to know do you have meetings where things related to ONNX-MLIR are discussed. It will be helpful to know future plans.

Thank you for offering to lower Shape.
Currently, we do not have a public meeting for ONNX-MLIR discussion. As the participants are increasing, we should consider that. Good point!

File: test/mlir/onnx/onnx_lowering.mlir (hunk 2 above)
Comment (tungld) [reply test/mlir/onnx/onnx_lowering.mlir:2153]: LGTM! Thanks!

**Comment by pashu123** [comment]
@chentong319 , @tungld  Could you merge this PR?

**Comment by tjingrant** [comment]
@jenkins-droid test this please.

**Comment by tjingrant** [comment]
I'm to blame for the unit tests. I need to set the types of output tensors in python runtime correctly.

**Comment by tjingrant** [comment]
@jenkins-droid test this please.

**Pull request merged**